Merge otherlibs/dynlink/Makefile into the root Makefile

shindere opened the pull request:
This PR continues the build system unification work started in #11243, #11248, #11268, #11420, #11675, #12198, #12321, #12586, #12616, #12706 and #13048.

The present PR merges `otherlibs/dynlink/Makefile` into the root `Makefile` and consists of four commits. The three first ones have the property of not altering the installed artefacts at all. The only commit that does so is the last one, which is the core of the PR, and the only thing it changes to the installed artifacts is the checksums, which is due to the fact that the compilation of the OCaml sources is done from the top-level of the source tree rather than from the `otherlibs/dynlink` directory as before.

The first commit introduces the private `OTHERLIBS` build variable to keep track of which other libraries are still built by `otherlibs/Makefile`. This variable is initially equal to `OTHERLIBRARIES` (the list of other libraries to build) but this list will become shorter andshohrter as more and more other libraries will be built directly by the root Makefile. Once all the other libraries are built by the root Makefile this variable is expected to become empty and to finally vanish.

The second commit delegates some file copying which was previously done during the build to the configure script.

THe third (and last preparatory) commit extends the general framework defined in `Makefile.common` to make sure the `.cmx` files of a native library get copied to the directory where its `.cmxa` file resides, as this is necessary when linking with dynlink's native version. As the copy happens also for all the other libraries built by the framework, the `clean` rule of `compilerlibs` had to be adjusted to clean the newly copied artefacts.

Finally, the PR's last commit is its core, which is the actual merge of `otherlibs/dynlink/Makefile` into the root `Makefile`. As usually, this goes hand in hand with `otherlibs/dynlink/.depend` being merged into the root `.depend`.

shindere committed:
Build system: introduce the private OTHERLIBS variable

This variable lists the other libraries that get built by the
recursive makefiles.

At this point it has the same content as the OTHERLIBRARIES variable
but, when the other libraries will be built by the root Makefile
their names will be removed from this variable.

It is thus expected that this variable becomes empty and can be removed
once all the other libraries get built by the root Makefile.

dustanddreams commented on Makefile:
```diff
@@ -1559,7 +1586,13 @@ clean::
 
-subdirs = stdlib $(addprefix otherlibs/, $(ALL_OTHERLIBS))
+# The following definition duplicates the otherlibs/ prefix but this
+# will go away with the merge of the sub  makefiles
```
```suggestion
# will go away with the merge of the sub makefiles.
```

shindere commented:
Fixed, thanks!

shindere committed:
Testsuite: simplify the C# test: do not spell-out dynlink libraries

Before this commit, the test was adding dynlink.cmx?a to the
all_modules variable but this is actually not necessary because
the "include dynlink" directive already adds them as appropriate to
the libraries variable.

This has been spotted by @dra27 as it led to a test failure when
-linkall was (incorrectly) used because this led to some modules
being linked twice.

shindere committed:
Build system: introduce per-program variables to store link flags

This commit introduces build variables

  foo_{COMMON,BYTECODE,NATIVE}_LINKFLAGS

where foo is the name of a program to build.

These variables are initialised with empty strings (no flags) and
can then be overridden by each program which needs to.

Such variables allow more fine-grained control over flags so that they
do not propagate when put in target-specific variables.

It may be the case that these variables become unnecessary once we
use a version of GNU make which is recent enough to support the
private keyword.

shindere commented:
A few updates:

The PR is now ready to be reviewed, after a bit of work,
part of which having been done offline with @dra27.

To elaborate on the change of checksums, this is actually due
to the fact that this PR changes the file names that are
stored in locations. Instead of having jsut file names as we
did before, we now have complete file paths relative to the
top level directory of the source tree. This can be seen in the
changes to the test reference files.

Also, thePR now begins with a commit that simplifies the `C#` test
by not mentionning the dynlink files explicitly. THey are actually
already linked so they were linked twice, which caused a problem when used
in conjunction with `-linkall`. This is @dra27's discovery.

The commit which had been added to copy `.cmx` files systematically has
finally been remove in favour of a more ad hoc solution, which is closer
tohwat was done before.

shindere committed:
Let configure generate otherlibs/dynlink/dynlink_platform_intf.mli

otherlibs/dynlink/dynlink_platform_intf.mli is a copy of the
corresponding otherlibs/dynlink/dynlink_platform_intf.ml file.
Before this commit, the copy was done during the build. With this commit
the copy (more likely a symlink) is done by configure as is already
done for other files that need to be duplicated.

shindere committed:
Makefile: remove duplicate target-specific variable definition

shindere committed:
Merge otherlibs/dynlink/Makefile into the root Makefile

dra27 reviewed (approved):
Reviewed offline (outcome in the most recent force push) - LGTM, thanks, @shindere

shindere commented:
Plenty of thanks @ëra27 for your super supportive review.
